pkg/util/camel/catalog.go

```diff
@@ -91,13 +93,110 @@ func (c Catalog) VisitArtifacts(visitor func(string, Artifact) bool) {
 }
 
 // VisitSchemes --
-func (c Catalog) VisitSchemes(visitor func(string, Scheme) bool) {
+func (c RuntimeCatalog) VisitSchemes(visitor func(string, Scheme) bool) {
 	for id, scheme := range c.schemesByID {
 		if !visitor(id, scheme) {
 			break
 		}
 	}
 }
 
+// ******************************
+//
+//
+//
+// ******************************
+
+var defaultCatalog RuntimeCatalog
+var catalogs map[string]RuntimeCatalog
+var catalogsLock sync.Mutex
+
+func init() {
+	c, err := loadCatalog("camel-catalog.yaml")
+	if err != nil {
+		panic(err)
+	}
+
+	defaultCatalog = *c
+
+	catalogs = make(map[string]RuntimeCatalog)
+}
+
+func loadCatalog(resourceName string) (*RuntimeCatalog, error) {
+	var catalog Catalog
+
+	data, ok := deploy.Resources[resourceName]
+	if !ok {
+		return nil, nil
+	}
+
+	if err := yaml.Unmarshal([]byte(data), &catalog); err != nil {
+		return nil, err
+	}
+
+	rc := RuntimeCatalog{}
+	rc.Catalog = catalog
+	rc.artifactByScheme = make(map[string]string)
+	rc.schemesByID = make(map[string]Scheme)
+
+	for id, artifact := range catalog.Artifacts {
+		for _, scheme := range artifact.Schemes {
+			scheme := scheme
+			rc.artifactByScheme[scheme.ID] = id
+			rc.schemesByID[scheme.ID] = scheme
+		}
+	}
+
+	return &rc, nil
+}
+
 // Runtime --
-var Runtime Catalog
+func Runtime(camelVersion string) *RuntimeCatalog {
```

Comment:
I wonder whether we should rename the method to something like `RuntimeCatalog` or `Catalog`?

Reply:
renamed to `func Catalog(...) *RuntimeCatalog`